src/allocator.rs

```diff
@@ -25,30 +26,42 @@ enum ObjectType {
 impl NodePtr {
     pub const NIL: Self = Self::new(ObjectType::SmallAtom, 0);
 
-    const fn new(t: ObjectType, idx: usize) -> Self {
-        debug_assert!(idx <= NODE_PTR_IDX_MASK as usize);
-        NodePtr(((t as u32) << NODE_PTR_IDX_BITS) | (idx as u32))
+    const fn new(object_type: ObjectType, index: usize) -> Self {
+        debug_assert!(index <= NODE_PTR_IDX_MASK as usize);
+        NodePtr(((object_type as u32) << NODE_PTR_IDX_BITS) | (index as u32))
     }
 
-    fn node_type(&self) -> (ObjectType, u32) {
-        (
-            match self.0 >> NODE_PTR_IDX_BITS {
-                0 => ObjectType::Pair,
-                1 => ObjectType::Bytes,
-                2 => ObjectType::SmallAtom,
-                _ => {
-                    panic!("unknown NodePtr type");
-                }
-            },
-            (self.0 & NODE_PTR_IDX_MASK),
+    pub fn is_atom(self) -> bool {
+        matches!(
+            self.object_type(),
+            ObjectType::Bytes | ObjectType::SmallAtom
         )
     }
 
-    pub(crate) fn as_index(&self) -> usize {
-        match self.node_type() {
-            (ObjectType::Pair, idx) => (idx as usize) * 3,
-            (ObjectType::Bytes, idx) => (idx as usize) * 3 + 1,
-            (ObjectType::SmallAtom, idx) => (idx as usize) * 3 + 2,
+    pub fn is_pair(self) -> bool {
+        self.object_type() == ObjectType::Pair
+    }
+
+    fn object_type(self) -> ObjectType {
+        match self.0 >> NODE_PTR_IDX_BITS {
+            0 => ObjectType::Pair,
+            1 => ObjectType::Bytes,
+            2 => ObjectType::SmallAtom,
+            _ => unreachable!(),
+        }
+    }
+
+    fn index(self) -> u32 {
+        self.0 & NODE_PTR_IDX_MASK
+    }
+
+    pub(crate) fn as_unique_index(self) -> usize {
```

Comment:
I think we need to remove this function eventually. The one place that uses it should use a `HashMap` instead.

Reply:
I'll do this in another PR